File: meeting-rooms.py
class Solution:
# test cases:
#     input:        [[0,30],[5,10],[15,20]]
# after sorting:    [[5,10],[15,20],[0,30]]
# checking overlaps:[[5,10],[15,20],[0,30]]
# => false

# input: [[0,5]]
# => true

    def canAttendMeetings(self, intervals: List[List[int]]) -> bool:

        # sort by each element's 1th element
        intervals.sort(key = lambda x:x[1])

        # check that there are no overlaps between a given interval's starting and previous interval's ending time
        for i in range(1, len(intervals)):
            if intervals[i][0] < intervals[i - 1][1]:
                return False

        return True

File: cpp/meeting_rooms.cpp
class Solution {
public:
    // optimal sorting approach
    bool canAttendMeetings(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), [](auto &interval, auto &other) {
            return interval[0] < other[0];
        });

        for (int i = 1; i < intervals.size(); ++i) {
            // found a meeting that ends after the next one starts (overlap)
            if (intervals[i - 1][1] > intervals[i][0]) {
                return false;
            }
        }

        return true;
    }
};
